planner/core/rule_join_reorder_greedy.go

```diff
@@ -51,10 +55,62 @@ func (s *joinReorderGreedySolver) solve(joinNodePlans []LogicalPlan) (LogicalPla
 			cumCost: s.baseNodeCumCost(node),
 		})
 	}
+
+	for _, edge := range s.directedEdges {
+		nodeSet := s.directMap[edge[0]]
+		if nodeSet == nil {
+			nodeSet = make(map[LogicalPlan]struct{})
+			s.directMap[edge[0]] = nodeSet
+		}
+		nodeSet[edge[1]] = struct{}{}
+	}
+
+	isChanged := true
+	for isChanged {
+		isChanged = false
+		for _, set := range s.directMap {
+			oldSize := len(set)
+			newSet := make(map[LogicalPlan]struct{})
+			for k1 := range set {
+				for k2 := range s.directMap[k1] {
+					newSet[k2] = struct{}{}
+				}
+			}
+
+			for k2 := range newSet {
+				set[k2] = struct{}{}
+			}
+			newSize := len(set)
+			if oldSize != newSize {
+				isChanged = true
+			}
+		}
+	}
+
+	// First sort plans by cost
 	sort.SliceStable(s.curJoinGroup, func(i, j int) bool {
 		return s.curJoinGroup[i].cumCost < s.curJoinGroup[j].cumCost
 	})
 
+	// Then adjust the order for the plans according to join order that maintained in the directMap.
+	for key, set := range s.directMap {
+		keyIdx := -1
+		lowerStart := -1
+		for idx, node := range s.curJoinGroup {
+			if key == node.p {
+				keyIdx = idx
+			}
+			if _, ok := set[node.p]; ok && lowerStart < 0 {
+				lowerStart = idx
+			}
+		}
+		if keyIdx > lowerStart {
+			selectNode := s.curJoinGroup[keyIdx]
+			s.curJoinGroup = append(s.curJoinGroup[0:keyIdx], s.curJoinGroup[keyIdx+1:]...)
+			s.curJoinGroup = append(s.curJoinGroup[0:lowerStart], append([]*jrNode{selectNode}, s.curJoinGroup[lowerStart:]...)...)
+		}
+	}
+
```

Comment:
I think we can merge this part into the sort.
Since the direct edge actually is a partial-order property.